pkg/edgecache/finalizer/finalizer.go

```diff
@@ -0,0 +1,122 @@
+	"sigs.k8s.io/blob-csi-driver/pkg/util"
+)
+
+func GetPVByVolumeID(client clientset.Interface, volumeID string) (*v1.PersistentVolume, error) {
+	klog.Infof("No pvName provided, looking up via volumeID: %s", volumeID)
+	pvList, err := client.CoreV1().PersistentVolumes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Infof("unable to list volumes via volumeID: %s", volumeID)
+		return nil, err
+	}
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI.VolumeHandle == volumeID {
+			return &pv, nil
+		}
+	}
+	return nil, errors.New("no pv found")
+}
+
+func GetPVByName(client clientset.Interface, pvName string) (*v1.PersistentVolume, error) {
+	pv, err := client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("Unable to get PV %s", pvName)
+		return nil, err
+	}
+	return pv, nil
+}
+
+func AddFinalizer(client clientset.Interface, pv *v1.PersistentVolume, storageAccount string, containerName string) error {
+	/* add to claim first, then pv
+	We need on claim for dynamic pvc. deleting a pvc will trigger deleteVolume which then will delete az container
+	so intercept pvc deletes
+	*/
+	klog.V(3).Infof("AddFinalizer called with pv: %s, account: %s, container: %s", pv, storageAccount, containerName)
+	if util.ContainsString(pv.GetFinalizers(), finalizerName, nil) {
+		// finalizer already present means we succeeded here already
+		return nil
+	}
+	namespace := pv.Spec.ClaimRef.Namespace
+	pvc, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(context.TODO(), pv.Spec.ClaimRef.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("Unable to get PVC for pv %s", pv.Name)
+		return err
+	}
+	if !util.ContainsString(pvc.GetFinalizers(), finalizerName, nil) {
+		pvcClone := pvc.DeepCopy()
+		pvcClone.ObjectMeta.Finalizers = append(pvcClone.ObjectMeta.Finalizers, finalizerName)
+		_, err := client.CoreV1().PersistentVolumeClaims(namespace).Update(context.TODO(), pvcClone, metav1.UpdateOptions{})
+		if err != nil {
+			klog.V(3).Infof("Error adding protection finalizer to PVC %s: %v", pv.Name, err)
+			return err
+		}
+		klog.V(3).Infof("Added protection finalizer to PVC %s", pvc.Name)
+	} else {
+		klog.Infof("pvc already has finalizer: %v", pvc.GetFinalizers())
+	}
+	/* finally, update pv */
+	pvClone := pv.DeepCopy()
+	pvClone.ObjectMeta.Finalizers = append(pvClone.ObjectMeta.Finalizers, finalizerName)
+	pvClone.ObjectMeta.Annotations[accountAnnotation] = storageAccount
+	pvClone.ObjectMeta.Annotations[containerAnnotation] = containerName
+	_, err = client.CoreV1().PersistentVolumes().Update(context.TODO(), pvClone, metav1.UpdateOptions{})
+	if err != nil {
+		klog.V(3).Infof("Error adding protection finalizer and annotations to PV %s: %v", pv.Name, err)
+		return err
+	}
+	klog.V(3).Infof("Updated PV %s", pv.Name)
+	return nil
+}
+
+func RemoveFinalizer(client clientset.Interface, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) error {
```

Comment:
Do you need to remove the annotations here too? Or we assume they are benign if the delete fails downstream?

Reply:
i don't think it is necessary to remove them. annotations are ignored outside of the owner of annotation